File: src/modules/mail/mail-service.ts
import { injectable, inject } from "@needle-di/core";
import { and, desc, eq, sql } from "drizzle-orm";
import { DatabaseService } from "../../core/services/database-service.ts";
import { charactersTable, gmMailTable, mailTable } from "../../db/schema.ts";
import type { Mail } from "../../db/schema.ts";

/** One listed letter, as the mail handlers render it. */
export interface MailEntry {
  id: number;
  /** Position within the queried list — the wire's per-category index. */
  index: number;
  /** The other end of the letter: sender for the inbox, recipient for Sent. */
  counterparty: string;
  subject: string;
  body: string;
  /** Unix seconds — the wire's u32 timestamp. */
  sentAtEpoch: number;
  read: boolean;
  /** A letter with no sender character is from the system (Announcements). */
  systemSender: boolean;
}

/** Result of a fan-out send, per recipient — drives the 0x4801 failure list. */
export interface RecipientOutcome {
  name: string;
  /** undefined = delivered; otherwise the official failure code (-801/-802). */
  failureCode?: number;
}

/** The client's own per-category cap; a 17th inbox letter would silently vanish. */
export const MAILBOX_MAX = 16;

@injectable()
export class MailService {
  constructor(private readonly dbs = inject(DatabaseService)) {}

  private get db() {
    return this.dbs.get();
  }

  /** Delivers one letter to a named recipient. Returns the outcome row. */
  async send(
    senderCharacterId: number,
    senderName: string,
    recipientName: string,
    subject: string,
    body: string,
  ): Promise<RecipientOutcome> {
    const matches = await this.db
      .select({ id: charactersTable.id })
      .from(charactersTable)
      .where(eq(charactersTable.name, recipientName))
      .limit(1);

    const target = matches[0];
    if (!target) {
      return { name: recipientName, failureCode: -801 }; // RECIPIENT_UNKNOWN
    }

    const size = await this.mailboxSize(target.id);
    if (size >= MAILBOX_MAX) {
      return { name: recipientName, failureCode: -802 }; // RECIPIENT_MAILBOX_FULL
    }

    await this.db.insert(mailTable).values({
      sender_character_id: senderCharacterId,
      recipient_character_id: target.id,
      sender_name: senderName,
      recipient_name: recipientName,
      subject,
      body,
    });
    return { name: recipientName };
  }

  /** Undeleted inbox size — the send-side refusal check. */
  async mailboxSize(characterId: number): Promise<number> {
    const rows = await this.db
      .select({ count: sql<number>`count(*)::int` })
      .from(mailTable)
      .where(
        and(
          eq(mailTable.recipient_character_id, characterId),
          eq(mailTable.recipient_deleted, false),
        ),
      );
    return rows[0]?.count ?? 0;
  }

  /** Mail addressed to the Game Master (destination byte 3) — admin inbox only. */
  async sendGameMasterMail(
    senderCharacterId: number,
    senderName: string,
    subject: string,
    body: string,
  ): Promise<void> {
    await this.db.insert(gmMailTable).values({
      sender_character_id: senderCharacterId,
      sender_name: senderName,
      subject,
      body,
    });
  }

  /** The recipient's inbox, newest first, indexed 0..n. */
  async mailbox(characterId: number, limit = MAILBOX_MAX): Promise<MailEntry[]> {
    const rows = await this.db
      .select()
      .from(mailTable)
      .where(
        and(
          eq(mailTable.recipient_character_id, characterId),
          eq(mailTable.recipient_deleted, false),
        ),
      )
      .orderBy(desc(mailTable.sent_at), desc(mailTable.id))
      .limit(limit);
    return rows.map((row, index) => toEntry(row, index, row.sender_name, false));
  }

  /** The sender's Sent list, newest first, indexed 0..n. */
  async sentMail(characterId: number, limit = MAILBOX_MAX): Promise<MailEntry[]> {
    const rows = await this.db
      .select()
      .from(mailTable)
      .where(
        and(
          eq(mailTable.sender_character_id, characterId),
          eq(mailTable.sender_deleted, false),
        ),
      )
      .orderBy(desc(mailTable.sent_at), desc(mailTable.id))
      .limit(limit);
    return rows.map((row, index) => toEntry(row, index, row.recipient_name, true));
  }

  /** Opening is the only mark-as-read signal the protocol has — per side. */
  async markRead(mailId: number, sent: boolean): Promise<void> {
    await this.db
      .update(mailTable)
      .set(sent ? { sender_read: true } : { recipient_read: true })
      .where(eq(mailTable.id, mailId));
  }

  /** The letter leaves the asking end's list only — one row, two flags. */
  async delete(mailId: number, sent: boolean): Promise<void> {
    await this.db
      .update(mailTable)
      .set(sent ? { sender_deleted: true } : { recipient_deleted: true })
      .where(eq(mailTable.id, mailId));
  }

  /** Whether the GM mailbox has any letters (used by tests/tooling). */
  async gmMailCount(): Promise<number> {
    const rows = await this.db
      .select({ count: sql<number>`count(*)::int` })
      .from(gmMailTable);
    return rows[0]?.count ?? 0;
  }
}

function toEntry(row: Mail, index: number, counterparty: string, sent: boolean): MailEntry {
  return {
    id: row.id,
    index,
    counterparty,
    subject: row.subject,
    body: row.body,
    sentAtEpoch: Math.floor((row.sent_at?.getTime() ?? 0) / 1000),
    read: sent ? row.sender_read : row.recipient_read,
    systemSender: row.sender_character_id === null,
  };
}

File: src/Shared/Domain/Mail/MailService.cs
using Mgo2Server.Shared.Persistence;
using Mgo2Server.Shared.Persistence.Entities;
using Microsoft.EntityFrameworkCore;

namespace Mgo2Server.Shared.Domain.Mail;

/// <summary>One listed letter, as the mail handlers render it.</summary>
/// <param name="Identifier">Identifier of the letter.</param>
/// <param name="Index">Position within the queried list, the per-category wire index.</param>
/// <param name="Counterparty">The other end of the letter: sender for the inbox, recipient for Sent.</param>
/// <param name="Subject">Subject line of the letter.</param>
/// <param name="Body">Body of the letter.</param>
/// <param name="SentAtEpoch">Unix seconds the letter was sent at.</param>
/// <param name="Read">Whether this end has read the letter.</param>
/// <param name="SystemSender">Whether the letter was sent by the system.</param>
public sealed record MailEntry(
    int Identifier,
    int Index,
    string Counterparty,
    string Subject,
    string Body,
    long SentAtEpoch,
    bool Read,
    bool SystemSender);

/// <summary>Result of a delivery attempt, per recipient.</summary>
/// <param name="Name">Name of the recipient.</param>
/// <param name="FailureCode">Official failure code, or <c>null</c> when the letter was delivered.</param>
public sealed record RecipientOutcome(string Name, int? FailureCode);

/// <summary>
/// Owns the mail. The recipient's inbox and the sender's sent list are the same
/// row, split by per-side read and deleted flags.
/// </summary>
/// <param name="contextFactory">Factory used to create database contexts.</param>
public sealed class MailService(IDbContextFactory<Mgo2DatabaseContext> contextFactory)
    : DomainService(contextFactory)
{
    /// <summary>The client's own per-category cap; a seventeenth letter would silently vanish.</summary>
    public const int MailboxMaximum = 16;

    /// <summary>Official failure code for an unknown recipient (-801).</summary>
    private const int RecipientUnknownFailureCode = -801;

    /// <summary>Official failure code for a full mailbox (-802).</summary>
    private const int RecipientMailboxFullFailureCode = -802;

    /// <summary>Delivers one letter to a named recipient.</summary>
    /// <param name="senderCharacterIdentifier">Character that sends the letter.</param>
    /// <param name="senderName">Name of the sender as shown in the inbox.</param>
    /// <param name="recipientName">Name of the recipient.</param>
    /// <param name="subject">Subject line of the letter.</param>
    /// <param name="body">Body of the letter.</param>
    /// <param name="cancellationToken">Token that cancels the operation.</param>
    public async Task<RecipientOutcome> SendAsync(
        int senderCharacterIdentifier,
        string senderName,
        string recipientName,
        string subject,
        string body,
        CancellationToken cancellationToken = default)
    {
        await using var context = await CreateContextAsync(cancellationToken);

        var targetIdentifier = await context.Characters
            .AsNoTracking()
            .Where(character => character.Name == recipientName)
            .Select(character => (int?)character.Identifier)
            .FirstOrDefaultAsync(cancellationToken);

        if (targetIdentifier is null)
        {
            return new RecipientOutcome(recipientName, RecipientUnknownFailureCode);
        }

        // Serialize deliveries to the same mailbox on its character row, so two
        // simultaneous senders cannot both see room and overshoot the cap.
        await using var transaction = await context.Database.BeginTransactionAsync(cancellationToken);
        await context.Database.ExecuteSqlAsync(
            $"SELECT id FROM characters WHERE id = {targetIdentifier.Value} FOR UPDATE",
            cancellationToken);

        var size = await MailboxSizeAsync(context, targetIdentifier.Value, cancellationToken);
        if (size >= MailboxMaximum)
        {
            await transaction.RollbackAsync(cancellationToken);
            return new RecipientOutcome(recipientName, RecipientMailboxFullFailureCode);
        }

        context.MailMessages.Add(new MailMessage
        {
            SenderCharacterIdentifier = senderCharacterIdentifier,
            RecipientCharacterIdentifier = targetIdentifier.Value,
            SenderName = senderName,
            RecipientName = recipientName,
            Subject = subject,
            Body = body,
        });

        await context.SaveChangesAsync(cancellationToken);
        await transaction.CommitAsync(cancellationToken);
        return new RecipientOutcome(recipientName, null);
    }

    /// <summary>Stores a letter addressed to the game masters.</summary>
    /// <param name="senderCharacterIdentifier">Character that sends the letter.</param>
    /// <param name="senderName">Name of the sender.</param>
    /// <param name="subject">Subject line of the letter.</param>
    /// <param name="body">Body of the letter.</param>
    /// <param name="cancellationToken">Token that cancels the operation.</param>
    public async Task SendGameMasterMailAsync(
        int senderCharacterIdentifier,
        string senderName,
        string subject,
        string body,
        CancellationToken cancellationToken = default)
    {
        await using var context = await CreateContextAsync(cancellationToken);
        context.GameMasterMail.Add(new GameMasterMail
        {
            SenderCharacterIdentifier = senderCharacterIdentifier,
            SenderName = senderName,
            Subject = subject,
            Body = body,
        });

        await context.SaveChangesAsync(cancellationToken);
    }

    /// <summary>Returns the undeleted inbox size of a character.</summary>
    /// <param name="characterIdentifier">Character whose mailbox is measured.</param>
    /// <param name="cancellationToken">Token that cancels the operation.</param>
    public async Task<int> GetMailboxSizeAsync(
        int characterIdentifier,
        CancellationToken cancellationToken = default)
    {
        await using var context = await CreateContextAsync(cancellationToken);
        return await MailboxSizeAsync(context, characterIdentifier, cancellationToken);
    }

    /// <summary>Returns the inbox of a character, newest first, indexed from zero.</summary>
    /// <param name="characterIdentifier">Character whose inbox is listed.</param>
    /// <param name="limit">Maximum number of letters to return.</param>
    /// <param name="cancellationToken">Token that cancels the operation.</param>
    public async Task<List<MailEntry>> GetMailboxAsync(
        int characterIdentifier,
        int limit = MailboxMaximum,
        CancellationToken cancellationToken = default)
    {
        await using var context = await CreateContextAsync(cancellationToken);
        var rows = await context.MailMessages
            .AsNoTracking()
            .Where(message => message.RecipientCharacterIdentifier == characterIdentifier &&
                !message.RecipientDeleted)
            .OrderByDescending(message => message.SentAt)
            .ThenByDescending(message => message.Identifier)
            .Take(limit)
            .ToListAsync(cancellationToken);

        return [.. rows.Select((row, index) => ToEntry(row, index, row.SenderName, sent: false))];
    }

    /// <summary>Returns the sent list of a character, newest first, indexed from zero.</summary>
    /// <param name="characterIdentifier">Character whose sent list is listed.</param>
    /// <param name="limit">Maximum number of letters to return.</param>
    /// <param name="cancellationToken">Token that cancels the operation.</param>
    public async Task<List<MailEntry>> GetSentMailAsync(
        int characterIdentifier,
        int limit = MailboxMaximum,
        CancellationToken cancellationToken = default)
    {
        await using var context = await CreateContextAsync(cancellationToken);
        var rows = await context.MailMessages
            .AsNoTracking()
            .Where(message => message.SenderCharacterIdentifier == characterIdentifier &&
                !message.SenderDeleted)
            .OrderByDescending(message => message.SentAt)
            .ThenByDescending(message => message.Identifier)
            .Take(limit)
            .ToListAsync(cancellationToken);

        return [.. rows.Select((row, index) => ToEntry(row, index, row.RecipientName, sent: true))];
    }

    /// <summary>Marks one side of a letter as read.</summary>
    /// <param name="mailIdentifier">Identifier of the letter.</param>
    /// <param name="sent"><c>true</c> to mark the sender's side, <c>false</c> the recipient's.</param>
    /// <param name="cancellationToken">Token that cancels the operation.</param>
    public async Task MarkReadAsync(
        int mailIdentifier,
        bool sent,
        CancellationToken cancellationToken = default)
    {
        await using var context = await CreateContextAsync(cancellationToken);
        var query = context.MailMessages.Where(message => message.Identifier == mailIdentifier);

        if (sent)
        {
            await query.ExecuteUpdateAsync(
                setters => setters.SetProperty(message => message.SenderRead, true),
                cancellationToken);
        }
        else
        {
            await query.ExecuteUpdateAsync(
                setters => setters.SetProperty(message => message.RecipientRead, true),
                cancellationToken);
        }
    }

    /// <summary>Deletes one side of a letter.</summary>
    /// <param name="mailIdentifier">Identifier of the letter.</param>
    /// <param name="sent"><c>true</c> to delete from the sender's list, <c>false</c> the recipient's.</param>
    /// <param name="cancellationToken">Token that cancels the operation.</param>
    public async Task DeleteAsync(
        int mailIdentifier,
        bool sent,
        CancellationToken cancellationToken = default)
    {
        await using var context = await CreateContextAsync(cancellationToken);
        var query = context.MailMessages.Where(message => message.Identifier == mailIdentifier);

        if (sent)
        {
            await query.ExecuteUpdateAsync(
                setters => setters.SetProperty(message => message.SenderDeleted, true),
                cancellationToken);
        }
        else
        {
            await query.ExecuteUpdateAsync(
                setters => setters.SetProperty(message => message.RecipientDeleted, true),
                cancellationToken);
        }
    }

    /// <summary>Returns how many letters the game-master mailbox holds.</summary>
    /// <param name="cancellationToken">Token that cancels the operation.</param>
    public async Task<int> GetGameMasterMailCountAsync(CancellationToken cancellationToken = default)
    {
        await using var context = await CreateContextAsync(cancellationToken);
        return await context.GameMasterMail.CountAsync(cancellationToken);
    }

    private static Task<int> MailboxSizeAsync(
        Mgo2DatabaseContext context,
        int characterIdentifier,
        CancellationToken cancellationToken) =>
        context.MailMessages
            .AsNoTracking()
            .CountAsync(
                message => message.RecipientCharacterIdentifier == characterIdentifier &&
                    !message.RecipientDeleted,
                cancellationToken);

    private static MailEntry ToEntry(MailMessage row, int index, string counterparty, bool sent) =>
        new(
            row.Identifier,
            index,
            counterparty,
            row.Subject,
            row.Body,
            new DateTimeOffset(DateTime.SpecifyKind(row.SentAt, DateTimeKind.Utc)).ToUnixTimeSeconds(),
            sent ? row.SenderRead : row.RecipientRead,
            row.SenderCharacterIdentifier is null);
}
